python/pyspark/sql/_typing.pyi

```diff
@@ -54,4 +54,10 @@ class SupportsClose(Protocol):
     def close(self, error: Exception) -> None: ...
 
 class UserDefinedFunctionLike(Protocol):
-    def __call__(self, *_: ColumnOrName) -> Column: ...
+    func: Callable[..., Any]
+    evalType: int
+    deterministic: bool
+    @property
+    def returnType(self) -> pyspark.sql.types.DataType: ...
+    def __call__(self, *cols: ColumnOrName) -> Column: ...
```

Comment:
To be precise [wrapper uses](https://github.com/apache/spark/blob/2fe12a75206d4dbef6d7678b876c16876136cdd0/python/pyspark/sql/udf.py#L198) `*args` not `*cols`.  I used `_`, because what we advertise to the end user, is different from wrapped signature (we might be able to do better, once we drop support for Python < 3.10, but that's far away from now). If we want to change this, I'd recommend `*args` not `*cols`.

Other than that LGTM.

Reply:
Oh my bad, that's a good catch! However, if we change this to be `args` then a `UserDefinedFunction` wouldn't type properly as a `UserDefinedFunctionLike`, would it?

Reply:
If I recall correctly only keyword argument names in `Protocols` are checked (and this can be explicitly disabled)\, but even if it wasn't the case, I don't think our intention was ever to cover `UserDefinedFunction` instances.

Reply:
(Specifically, we needed a way to express a fact, that object that we get, is callable and takes variable number of `str` or `Column` arguments and returns a `Column`. This is something that cannot be done with simple `Callable`, but is quite useful for the end user)

Reply:
Thanks @luranhe ‒ I'll merge it tomorrow, unless there are any further comments.